### ISSUE TITLE
Not very convincing mnist result: 4% accuracy

### ISSUE THREAD

**Issue by IvanVolosyuk** [Artelnics/opennn#137]
I build openn and examples on Linux and ran mnist example. Output is below. Doesn't look convincing to me. As I just evaluate opennn for my project, this will be a strong no-go. Something is definetely broken somewhere.

OpenNN. MNIST Example.
Layers number: 9
Layer 1: Scaling
Layer 2: Convolutional
Layer 3: Pooling
Layer 4: Convolutional
Layer 5: Pooling
Layer 6: Convolutional
Layer 7: Pooling
Layer 8: Perceptron
Layer 9: Probabilistic
Training with stochastic gradient descent...
Epoch 0;
Parameters norm: 106.043
Training loss: 1.48704
Batch size: 5
Gradient norm: 0.0345143
Learning rate: 0.01
Elapsed time: 00:11
Selection error: 0
Epoch 1;
Parameters norm: 106.043
Training loss: 1.48671
Batch size: 5
Gradient norm: 0.0387483
Learning rate: 0.01
Elapsed time: 00:22
Selection error: 0
Epoch 2;
Parameters norm: 106.043
Training loss: 1.48639
Batch size: 5
Gradient norm: 0.0350705
Learning rate: 0.01
Elapsed time: 00:33
Selection error: 0
Epoch 3;
Parameters norm: 106.043
Training loss: 1.48604
Batch size: 5
Gradient norm: 0.016576
Learning rate: 0.01
Elapsed time: 00:45
Selection error: 0
Epoch 4;
Parameters norm: 106.043
Training loss: 1.48571
Batch size: 5
Gradient norm: 0.0102665
Learning rate: 0.01
Elapsed time: 00:56
Selection error: 0
Epoch 5;
Parameters norm: 106.043
Training loss: 1.48537
Batch size: 5
Gradient norm: 0.034326
Learning rate: 0.01
Elapsed time: 01:07
Selection error: 0
Epoch 6;
Parameters norm: 106.043
Training loss: 1.48503
Batch size: 5
Gradient norm: 0.0341277
Learning rate: 0.01
Elapsed time: 01:19
Selection error: 0
Epoch 7;
Parameters norm: 106.043
Training loss: 1.4847
Batch size: 5
Gradient norm: 0.0184346
Learning rate: 0.01
Elapsed time: 01:30
Selection error: 0
Epoch 8;
Parameters norm: 106.043
Training loss: 1.48435
Batch size: 5
Gradient norm: 0.0377195
Learning rate: 0.01
Elapsed time: 01:41
Selection error: 0
Epoch 9;
Parameters norm: 106.043
Training loss: 1.48403
Batch size: 5
Gradient norm: 0.0403547
Learning rate: 0.01
Elapsed time: 01:53
Selection error: 0
Epoch 10;
Parameters norm: 106.043
Training loss: 1.48368
Batch size: 5
Gradient norm: 0.0285667
Learning rate: 0.01
Elapsed time: 02:04
Selection error: 0
Epoch 11;
Parameters norm: 106.043
Training loss: 1.48334
Batch size: 5
Gradient norm: 0.0523294
Learning rate: 0.01
Elapsed time: 02:15
Selection error: 0
Epoch 12: Maximum number of iterations reached.
Parameters norm: 106.043
Training loss: 1.48299
Batch size: 5
Gradient norm: 0.0289222
Learning rate: 0.01
Elapsed time: 02:27
Selection error: 0

Confusion matrix:

1 0 0 0 0 0 0 2 0 0
0 0 0 0 0 0 1 1 1 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 1 1 1
0 0 0 0 0 1 0 1 0 0
0 0 0 0 0 0 0 0 0 0
4 0 0 0 0 0 0 1 0 0
0 0 0 0 0 0 0 0 0 1
1 0 1 0 1 0 0 0 0 0
0 0 2 0 1 0 2 0 0 0

Accuracy: 4 %

**Comment by shleym2000** [comment]
I noticed that it stopped training process at epoch 12 because 12 was set as limit.

I wonder if results would be better if limit on epochs is not set (or changed to some outlandish large number).

Other stoppage criteria could stop process when it is really done to max perfection

**Comment by IvanVolosyuk** [comment]
I might be used to different terminology, but I would expect accuracy to be 90%+.
I tried to tweak the training rate (0.1) and increase limit on the epochs. 0% accuracy means to me that the model produced 0 correct answers on the test data.

Logs:
Epoch 157: Maximum training time reached.
Parameters norm: 105.301
Training loss: 0.98241
Batch size: 5
Gradient norm: 0.0137382
Learning rate: 0.1
Elapsed time: 30:03
Selection error: 0


Confusion matrix:

0 0 0 1 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 3
1 0 0 0 0 0 0 1 0 0
1 0 0 0 0 0 3 0 0 1
1 0 0 0 0 1 0 0 0 2
0 0 0 0 0 0 0 0 0 1
0 0 0 0 0 0 0 0 0 0
1 0 0 0 3 0 0 0 1 1
1 0 0 1 0 0 1 0 0 0
0 0 0 0 0 0 0 0 0 0

Accuracy: 0 %

**Comment by croja02** [comment]
Has this ever been resolved?
On a similar platform (ubunto 18.04) I have the same issues. Even on the IRIS problem I get a confusion matrix that looks like:

Confusion: 
0 0 8
0 0 7
0 0 15

**Comment by jandres-artelnics** [comment]
Hi @IvanVolosyuk ,

Have you tried using dev? We're making major changes on dev branch.

![image](https://user-images.githubusercontent.com/59603057/91956522-0bfd7200-ed05-11ea-8681-a63f3f2f2441.png)

Here's a pic from MNIST example. 

As you expected, accuracy is over 90%

@croja02 and @IvanVolosyuk , I'll try it on ubuntu just in case there's something weird.

Thank you for your comments.

**Comment by jandres-artelnics** [comment]
Hi @croja02 and @IvanVolosyuk ,

Using dev branch IRIS example works fine on xubuntu.

![image](https://user-images.githubusercontent.com/59603057/92245296-71e42800-eec4-11ea-8c10-3c16aee16df1.png)

I close the issue. 

If there's something wrong please feel free to open another issue. 

Thank you for your comments and sorry for our late reply.
